src/awkward/contents/content.py

```diff
@@ -42,14 +42,45 @@ class Content:
     is_union = False
 
     def _init(self, parameters: dict[str, Any] | None, backend: Backend):
-        if parameters is not None and not isinstance(parameters, dict):
+        if parameters is None:
+            pass
+        elif not isinstance(parameters, dict):
             raise ak._errors.wrap_error(
                 TypeError(
                     "{} 'parameters' must be a dict or None, not {}".format(
                         type(self).__name__, repr(parameters)
                     )
                 )
             )
+        else:
+            if not self.is_list and parameters.get("__array__") in (
+                "string",
+                "bytestring",
+                "sorted_map",
+            ):
+                raise ak._errors.wrap_error(
+                    TypeError(
+                        '{} is not allowed to have parameters["__array__"] = "{}"'.format(
+                            type(self).__name__, parameters["__array__"]
+                        )
+                    )
+                )
+            if not self.is_numpy and parameters.get("__array__") in ("char", "byte"):
+                raise ak._errors.wrap_error(
+                    TypeError(
+                        '{} is not allowed to have parameters["__array__"] = "{}"'.format(
+                            type(self).__name__, parameters["__array__"]
+                        )
+                    )
+                )
+            if not self.is_indexed and parameters.get("__array__") == "categorical":
+                raise ak._errors.wrap_error(
+                    TypeError(
+                        '{} is not allowed to have parameters["__array__"] = "{}"'.format(
+                            type(self).__name__, parameters["__array__"]
+                        )
+                    )
+                )
```

Comment:
If a layout node type is _not_ supposed to have a certain parameter, then that is now enforced. Sometimes, `"string"` parameters got onto UnionArrays because parameters from various levels were over-merged.